src/txrequest.h

```diff
@@ -0,0 +1,342 @@
+ *   peer), with only AlreadyHaveTx and other RequestedTx calls (both for other txids) in between. Any other
+ *   non-const operation removes the ability to call RequestedTx.
+ */
+class TxRequestTracker {
+    //! Configuration option: delay before downloading from inbound connection.
+    const std::chrono::microseconds m_inbound_delay;
+    //! Configuration option: timeout after requesting a download.
+    const std::chrono::microseconds m_timeout;
+
+    //! The various states a (txid,node) pair can be in.
+    //! Note that CANDIDATE is split up into 3 substates (NEW, BEST, OTHER), allowing more efficient implementation.
+    //! Also note that the sorting order of EntryTxid relies on the specific order of values in this enum.
+    enum class State : uint8_t {
+        //! A CANDIDATE entry whose reqtime is in the future.
+        CANDIDATE_NEW,
+        //! The best CANDIDATE for a given txid; only if there is no REQUESTED entry already for that txid.
+        //! The CANDIDATE_BEST is the lowest-priority entry among all CANDIDATE_OTHER (and _BEST) ones for that txid.
+        CANDIDATE_BEST,
+        //! A REQUESTED entry.
+        REQUESTED,
+        //! A CANDIDATE entry that's not CANDIDATE_NEW or CANDIDATE_BEST.
+        CANDIDATE_OTHER,
+        //! A COMPLETED entry.
+        COMPLETED,
+    };
+
+    //! Tag for the EntryPeer-based index.
+    struct ByPeer {};
+    //! Tag for the EntryTxid-based index.
+    struct ByTxid {};
+    //! Tag for the EntryTime-based index.
+    struct ByTime {};
+
+    //! The ByPeer index is sorted by (peer, state == CANDIDATE_BEST, txid)
+    using EntryPeer = std::tuple<uint64_t, bool, const uint256&>;
+
+    //! The ByTxid index is sorted by (txid, state, priority [CANDIDATE_OTHER]; 0 [otherwise])
+    using EntryTxid = std::tuple<const uint256&, State, uint64_t>;
+
+    //! The ByTime index is sorted by (0 [CANDIDATE_NEW,REQUESTED]; 1 [COMPLETED];
+    // 2 [CANDIDATE_OTHER,CANDIDATE_BEST], time)
+    using EntryTime = std::pair<int, std::chrono::microseconds>;
+
+    //! The current sequence number. Increases for every announcement. This is used to sort txids returned by
+    //! GetRequestable in announcement order.
+    uint64_t m_sequence{0};
+
+    //! An announcement entry.
+    struct Entry {
+        //! Txid that was announced.
+        const uint256 m_txid;
+        //! For CANDIDATE_{NEW,BEST,OTHER} the reqtime; for REQUESTED the exptime
+        std::chrono::microseconds m_time;
+        //! What peer the request was from.
+        const uint64_t m_peer;
+        //! What sequence number this announcement has.
+        const uint64_t m_sequence : 59;
+        //! Whether it's from an outbound peer or not.
+        const bool m_inbound : 1;
+        //! What state this announcement is in
+        //! This is a uint8_t instead of a State to silence a GCC warning.
+        uint8_t m_state : 3;
```

Comment:
272a615fad6f93e1b7f9b6245ce9090d07fb4a51

No warnings are observed on GCC 9.3.0.
What GCC version you are referring to?

Reply:
See https://github.com/bitcoin/bitcoin/pull/19184#discussion_r444545289